### PR TITLE
Solutions to Chapter 2

### PR DESCRIPTION
**Pull request by DavSanchez**
### Solutions for Chapter 2

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -336,7 +359,9 @@ from it!
 ghci> :l src/Chapter2.hs
 -}
 subList :: Int -> Int -> [a] -> [a]
-subList = error "subList: Not implemented!"
+subList i1 i2 ls
+  | (i1 < 0) || (i2 < 0) = []
+  | otherwise = take (i2 - i1 + 1) $ drop i1 ls
```
Comment (chshersh) [inline src/Chapter2.hs]: Your implementation is correct when the right bound is lower than the left. But I think you can implement it more efficiently by adding a separate case to your check :slightly_smiling_face:

File: src/Chapter2.hs [hunk 2]
```diff
@@ -500,7 +526,9 @@ True
 >>> isThird42 [42, 42, 0, 42]
 False
 -}
-isThird42 = error "isThird42: Not implemented!"
+isThird42 :: (Eq a, Num a) => [a] -> Bool
+isThird42 (_:_:42:_) = True
+isThird42 _ = False
```
Comment (chshersh) [inline src/Chapter2.hs:532]: Well done on the polymorphic type signature and perfect patter :ok: :top:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -620,7 +649,11 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven (x:[]) = x : []
```
Comment (chshersh) [inline src/Chapter2.hs]: You can write this case a bit simpler :slightly_smiling_face: 
```suggestion
takeEven [x] = [x]
```

File: src/Chapter2.hs [hunk 4]
```diff
@@ -620,7 +649,11 @@ Write a function that takes elements of a list only on even positions.
...
+takeEven (x:_:[]) = x : []
```
Comment (chshersh) [inline src/Chapter2.hs]: This pattern is actually redundant. See, it has the same shape as the pattern on the next line. And when `xs` is empty list per the last case, function will call `takeEven` on an empty list and will return empty list as well. So you don't need to handle it separately. But it is very helpful for pattern-matching understanding to cover special and smaller cases first, so you can see how to generalize your pattern.

File: src/Chapter2.hs [hunk 5]
```diff
@@ -727,7 +760,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat $ map (\e -> replicate e e) l
```
Comment (chshersh) [inline src/Chapter2.hs]: Great! 👍🏼 

If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 6]
```diff
@@ -740,7 +773,8 @@ the list with only those lists that contain a passed element.
 
 🕯 HINT: Use the 'elem' function to check whether an element belongs to a list
 -}
-contains = error "contains: Not implemented!"
+contains :: Eq a => a -> [[a]] -> [[a]]
+contains n = filter (elem n)
```
Comment (chshersh) [inline src/Chapter2.hs:777]: Eta-reduction award :trophy:

File: src/Chapter2.hs [hunk 7]
```diff
@@ -841,7 +877,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n l
+  | n < 0 = []
+  | otherwise = take (length l) $ drop n $ cycle l
```
Comment (chshersh) [inline src/Chapter2.hs:884]: Unfortunately, `cycle` fails in runtime on empty lists :recycle: So you need to handle this case separately for this function to work properly.

File: src/Chapter2.hs [hunk 8]
```diff
@@ -857,7 +896,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind l = last l : rewind (init l)
```
Comment (chshersh) [inline src/Chapter2.hs]: Your solution is correct! However, in lists, it is quite slow to add anything at the end of the list, as well as taking elements from the list end. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Nice solutions! And well done on solving the Advanced Task :+1: 

I left some comments, suggestions and explanations as part of my feedback. I hope you find it useful :slightly_smiling_face:

File: src/Chapter2.hs (hunk 1 above)
Comment (DavSanchez) [reply src/Chapter2.hs]: You're totally right, thanks!

File: src/Chapter2.hs (hunk 5 above)
Comment (DavSanchez) [reply src/Chapter2.hs]: Great, thanks!

File: src/Chapter2.hs (hunk 7 above)
Comment (DavSanchez) [reply src/Chapter2.hs:884]: Aah I see. Thanks!

File: src/Chapter2.hs (hunk 8 above)
Comment (DavSanchez) [reply src/Chapter2.hs]: Aha, `go` patterns look awesome!

I think I got it right, can you check? Thanks!

File: src/Chapter2.hs (hunk 4 above)
Comment (DavSanchez) [reply src/Chapter2.hs]: I see, the thing was my implementation failed the first times and I supposed it had to do with patterns, so I went over-exhaustive and forgot to re-check 😄 Thanks!

**Comment by chshersh** [comment]
@DavSanchez Your fixes are great, and all solutions are correct :100: 
This PR is ready for merge :checkered_flag:

**Comment by DavSanchez** [comment]
> @DavSanchez Your fixes are great, and all solutions are correct :100: 
> 
> This PR is ready for merge :checkered_flag: 

Thanks a lot for your help @chshersh!

**Pull request merged**